Remove the Box from the ShapeHandle

Right now, the `ShapeHandle` is [defined](https://github.com/rustsim/ncollide/blob/master/src/shape/shape.rs#L164) as a `Arc<Box<Shape<N>>`. It would be more efficient to avoid the double allocation by defining it as `Arc<Shape<N>>` instead. This is not possible currently because we need to use `Arc::make_mut` at [some point](https://github.com/rustsim/ncollide/blob/master/src/shape/shape.rs#L179), which relies in the fact that `Box<Shape<N>>` implements `Clone`.

@Ralith suggested that we use [Arc::get_mut](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut) instead, after making sure we have a unique reference. This has the specificity of working on an `Arc<Shape<N>>`.

I made some implementation attempt. I first modified the `ShapeClone` trait to:

```rust
pub trait ShapeClone<N: RealField> {
    fn clone_arc(&self) -> Arc<Shape<N>> {
        unimplemented!()
    }
}

impl<N: RealField, T: 'static + Shape<N> + Clone> ShapeClone<N> for T {
    fn clone_arc(&self) -> Arc<Shape<N>> {
        Arc::new(self.clone())
    }
}
```

and the `ShapeHandle::make_mut` method to:

```rust
    pub(crate) fn make_mut(&mut self) -> &mut Shape<N> {
        if let Some(shape) = Arc::get_mut(&mut self.0) {
            return shape
        }

        let unique_self = self.0.clone_arc();
        self.0 = unique_self;
        Arc::get_mut(&mut self.0).unwrap()
    }
```

Unfortunately, this does not compiles because of borrowing issues:

```
error[E0502]: cannot borrow `self.0` as immutable because it is also borrowed as mutable
   --> build/ncollide3d/../../src/shape/shape.rs:161:27
    |
154 |     pub(crate) fn make_mut(&mut self) -> &mut Shape<N> {
    |                            - let's call the lifetime of this reference `'1`
155 |         {
156 |             if let Some(shape) = Arc::get_mut(&mut self.0) {
    |                                               ----------- mutable borrow occurs here
157 |                 return shape
    |                        ----- returning this value requires that `self.0` is borrowed for `'1`
...
161 |         let unique_self = self.0.clone_arc();
    |                           ^^^^^^ immutable borrow occurs here

error[E0506]: cannot assign to `self.0` because it is borrowed
   --> build/ncollide3d/../../src/shape/shape.rs:162:9
    |
154 |     pub(crate) fn make_mut(&mut self) -> &mut Shape<N> {
    |                            - let's call the lifetime of this reference `'1`
155 |         {
156 |             if let Some(shape) = Arc::get_mut(&mut self.0) {
    |                                               ----------- borrow of `self.0` occurs here
157 |                 return shape
    |                        ----- returning this value requires that `self.0` is borrowed for `'1`
...
162 |         self.0 = unique_self;
    |         ^^^^^^ assignment to borrowed `self.0` occurs here

error[E0499]: cannot borrow `self.0` as mutable more than once at a time
   --> build/ncollide3d/../../src/shape/shape.rs:163:22
    |
154 |     pub(crate) fn make_mut(&mut self) -> &mut Shape<N> {
    |                            - let's call the lifetime of this reference `'1`
155 |         {
156 |             if let Some(shape) = Arc::get_mut(&mut self.0) {
    |                                               ----------- first mutable borrow occurs here
157 |                 return shape
    |                        ----- returning this value requires that `self.0` is borrowed for `'1`
...
163 |         Arc::get_mut(&mut self.0).unwrap()
    |                      ^^^^^^^^^^^ second mutable borrow occurs here

error: aborting due to 3 previous errors
```

I'm not sure why this error appears.

--- a/src/shape/shape.rs
+++ b/src/shape/shape.rs
@@ -9,14 +9,13 @@ use std::sync::Arc;
 use downcast_rs::Downcast;
 
 pub trait ShapeClone<N: RealField> {
-    fn clone_box(&self) -> Box<dyn Shape<N>> {
-        unimplemented!()
-    }
+    /// Construct an `Arc` that refers to a uniquely-owned copy of `self`
+    fn clone_arc(&self) -> Arc<dyn Shape<N>>;
 }
 
 impl<N: RealField, T: 'static + Shape<N> + Clone> ShapeClone<N> for T {
-    fn clone_box(&self) -> Box<dyn Shape<N>> {
-        Box::new(self.clone())
+    fn clone_arc(&self) -> Arc<dyn Shape<N>> {
+        Arc::new(self.clone())
     }
 }
 
@@ -151,39 +150,37 @@ impl<N: RealField> dyn Shape<N> {
     }
 }
 
-impl<N: RealField> Clone for Box<dyn Shape<N>> {
-    fn clone(&self) -> Box<dyn Shape<N>> {
-        self.clone_box()
-    }
-}
-
 /// A shared handle to an abstract shape.
 ///
 /// This can be mutated using COW.
 #[derive(Clone)]
-pub struct ShapeHandle<N: RealField>(Arc<Box<dyn Shape<N>>>);
+pub struct ShapeHandle<N: RealField>(Arc<dyn Shape<N>>);
 
 impl<N: RealField> ShapeHandle<N> {
     /// Creates a sharable shape handle from a shape.
     #[inline]
     pub fn new<S: Shape<N>>(shape: S) -> ShapeHandle<N> {
-        ShapeHandle(Arc::new(Box::new(shape)))
-    }
-
-    /// Creates a sharable shape handle from a shape trait object.
-    pub fn from_box(shape: Box<dyn Shape<N>>) -> ShapeHandle<N> {
         ShapeHandle(Arc::new(shape))
     }
 
+    /// Creates a sharable shape handle from a shape trait object.
+    pub fn from_arc(shape: Arc<dyn Shape<N>>) -> ShapeHandle<N> {
+        ShapeHandle(shape)
+    }
+
     pub(crate) fn make_mut(&mut self) -> &mut dyn Shape<N> {
-        &mut **Arc::make_mut(&mut self.0)
+        if Arc::get_mut(&mut self.0).is_none() {
+            let unique_self = self.0.clone_arc();
+            self.0 = unique_self;
+        }
+        Arc::get_mut(&mut self.0).unwrap()
     }
 }
 
 impl<N: RealField> AsRef<dyn Shape<N>> for ShapeHandle<N> {
     #[inline]
     fn as_ref(&self) -> &dyn Shape<N> {
-        &*self.deref()
+        &*self.0
     }
 }
 
@@ -192,6 +189,6 @@ impl<N: RealField> Deref for ShapeHandle<N> {
 
     #[inline]
     fn deref(&self) -> &dyn Shape<N> {
-        &**self.0.deref()
+        &*self.0
     }
 }

